dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,407 @@
+  $ west init zephyr
+  $ cd zephyr
+  $ west update
+  ```
+
+- Install required packages, use virtual environment to avoid conflict with
+  user/system packages.
+
+  ```shell
+  $ python3 -m venv .env
+  $ pip3 install -r zephyr/scripts/requirements.txt
+  ```
+
+- Build Zephyr, we use Qemu as the target, but it works on APU (without DLME).
+
+  ```shell
+  $ cd zephyr
+  $ west build -p auto -b qemu_x86_64 samples/userspace/hello_world_user/
+  $ sudo cp build/zephyr/zephyr.elf /mnt/boot/
+  ```
+
+- Configure GRUB, create '/mnt/boot/grub/grub.cfg` with the following contents.
+
+  ```shell
+  default 0
+  menuentry "Zephyr (DLME)" {
+    slaunch skinit
+    slaunch_module /skl.bin
+    multiboot /zephyr.elf
+  }
+  menuentry "Zephyr (no DLME)" {
+    multiboot /zephyr.elf
+  }
+  ```
+
+- Umount disk and sync before unplugging it. There were some problems with
+  obtaining working binary, so before booting Zephyr on target platform please
+  check whether it boots in Qemu (without DLME). When booting Zephyr you should
+  see `Hello World` message on serial console, if Zephyr doesn't output anything
+  or if it stuck at `Booting Zephyr` message then something went wrong.
+
+  ```shell
+  $ qemu-system-x86_64 -nographic -serial mon:stdio -m 512M -cpu qemu64 \
+    -drive format=raw,file=/dev/disk/by-id/usb-TS-RDF5_SD_Transcend_000000000039-0\:0
+  ```
+
+- Zephyr should normally boot on APU. If DLME variant (from boot menu) is used
+  Zephyr will also boot normally. That is because TrenchBoot GRUB executes SKL
+  only when using
+  [Multiboot2](https://github.com/3mdeb/grub/blob/tb_xen/grub-core/loader/multiboot.c#L192)
+  but Zephyr has support only for Multiboot1.
+
+- Booting Zephyr in DLME would require updating `slaunch` support in GRUB and
+  bringing Multiboot1 support to SKL.
+
+## Summary
+
+* The above report outlines four operating systems that should be considered
+  candidates for Fobnail Token interoperability.
+
+* The basic choice is Linux, the operating system based on it is created with
+  the use of Yocto Project. All achievements can be reproduced at any time using
+  the [meta-fobnail](https://github.com/fobnail/meta-fobnail) layer.
+
+* As part of the report, the seL4, Xous and Zephyr systems were also checked.
+  The possibility of running it in DLME and the integration of the
+  [fobnail-attester](https://github.com/fobnail/fobnail-attester) application
+  was checked for each of them.
+
+* The description of an attempt to run Zephyr on PC Engines apu2 in order to
+  verify the current state of the system to work with Fobnail Token has been
+  included. The effects are described in the [PoC test](#poc-test) section.
+
+* The table below summarises all OSes features described
+  [above](#os-requirements).
+
+| OS      | USB host driver  | USB EEM driver   | Network stack | TPM driver        | OS portability | Bootloader capabilities | C library | Microkernel | CPU Architecture support | Bootable by SKL | Score |
```

Comment:
I don't know if comparing driver support for mix of microkernels and other types of kernels makes sense. By definition, some things that are drivers in bigger kernels are moved to userspace in microkernels, which may impact scores in this table.

Reply:
Bingo.  seL4 does not (and should not!) have USB support.  That’s the job of userspace.

Reply:
If we were very strict about apple to apple we will dive in scientific discussion. Extreme precision here would make whole decision process not feasible for size of project. Do you have alternative way to achieve quick summary, that can help make final decision?

Reply:
Comparison includes userspace drivers available for these microkernels. seL4 has no USB driver available, except this https://github.com/seL4/projects_libs/tree/master/libusbdrivers which is of unknown quality and has support for EHCI only.
Neither does CAmkES have USB drivers.

Reply:
If this is mentioned then yes. Although I am still surprised that USB support is so poor for those microkernels.

Reply:
@krystian-hebel I added mention, does it solve thread?

Reply:
Solved